libsolidity/formal/BMC.cpp

```diff
@@ -290,104 +291,214 @@ bool BMC::visit(Conditional const& _op)
+			!broke && !continues && expr(_node.condition()),
+			indicesAfter,
+			copyVariableIndices()
+		);
+		loopScopes.pop();
+		_node.condition().accept(*this);
+	}
 
 	m_loopExecutionHappened = true;
 	return false;
 }
 
-// Here we consider the execution of two branches similar to WhileStatement.
+// Unrolls for loop
 bool BMC::visit(ForStatement const& _node)
 {
 	if (_node.initializationExpression())
 		_node.initializationExpression()->accept(*this);
+	auto indicesBefore = copyVariableIndices();
 
-	auto indicesBeforeLoop = copyVariableIndices();
-
-	// Do not reset the init expression part.
 	auto touchedVars = touchedVariables(_node.body());
 	if (_node.condition())
 		touchedVars += touchedVariables(*_node.condition());
 	if (_node.loopExpression())
 		touchedVars += touchedVariables(*_node.loopExpression());
 
+	m_context.pushSolver();
+	// variables touched by loop might change their value
+	// assume values are not constant
 	m_context.resetVariables(touchedVars);
-
 	if (_node.condition())
 	{
 		_node.condition()->accept(*this);
-		if (isRootFunction())
+		// do not apply condition target on expressions
+		// which value might be changed by other loops
+		if (isRootFunction() && (!isInsideLoop()))
 			addVerificationTarget(
 				VerificationTargetType::ConstantCondition,
 				expr(*_node.condition()),
 				_node.condition()
 			);
 	}
-
-	m_context.pushSolver();
-	if (_node.condition())
-		m_context.addAssertion(expr(*_node.condition()));
-	_node.body().accept(*this);
-	if (_node.loopExpression())
-		_node.loopExpression()->accept(*this);
 	m_context.popSolver();
+	resetVariableIndices(indicesBefore);
 
-	auto indicesAfterLoop = copyVariableIndices();
-	// We reset the execution to before the loop
-	// and visit the condition.
-	resetVariableIndices(indicesBeforeLoop);
-	if (_node.condition())
-		_node.condition()->accept(*this);
+	smtutil::Expression broke(false);
+	smtutil::Expression forCondition(true);
+	unsigned int bmcLoopIterations = m_settings.bmcLoopIterations.value_or(1);
+	for (unsigned int i = 0; i < bmcLoopIterations; ++i)
+	{
+		if (_node.condition())
+		{
+			_node.condition()->accept(*this);
+			// values in loop condition might change during loop iteration
+			forCondition = expr(*_node.condition());
+		}
+		loopScopes.emplace();
+
+		auto indicesBefore = copyVariableIndices();
+		pushPathCondition(forCondition);
+		_node.body().accept(*this);
+		popPathCondition();
+
+		smtutil::Expression continues(false);
+		smtutil::Expression brokeInCurrentIteration(false);
+		for (auto const& loopControl: loopScopes.top())
+		{
+			if (loopControl.kind == LoopControlKind::Break)
+			{
+				// use SSAs associated with this break statement only if
+				// loop didn't break or continue earlier in the iteration
+				// loop condition is included in break path conditions
+				mergeVariables(
+					!brokeInCurrentIteration && !continues && loopControl.pathConditions,
+					loopControl.variableIndicies,
+					copyVariableIndices()
```

Comment:
I don't think this is correct. In case it didn't break and didn't continue, this SSA set is missing the `loopExpression()` potential changes, isn't it?

Reply:
Loop expression is accepted a couple lines lower in the code.
Actually there is a different problem - loop expression is executed even after `break` triggered.